### PR TITLE
Simplify install

### PR DESCRIPTION
**Pull request by Christian-Schl**
This Pull Request make it possible to use the pkg to create a single executable.

**Benefits**
- no node installation needed
- easy update process

The Data Dir with the Configurations can be controlled with the env var "ZIGBEE2MQTT_DATA" if this is not set the Config Dir is just the dir where the exe is located.

**How to create the executeable**

Execute 'npm run build-bin' 
the executable files will create in bin dir

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -1,4 +1,13 @@
 #!/usr/bin/env node
 const path = require('path');
-process.env['ZIGBEE2MQTT_DATA'] = process.env['ZIGBEE2MQTT_DATA'] || path.join(process.env['HOME'], '.z2m');
+
```
Comment (sjorge) [inline cli.js:3]: It’s cleaner to if this ENV var is not set, and then do the fallback inside that.

Comment (Christian-Schl) [reply cli.js:3]: Done

**Comment by Koenkk** [comment]
Could you also make a pr for the documentation? https://github.com/koenkk/zigbee2mqtt.io

File: cli.js [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+if (!process.env['ZIGBEE2MQTT_DATA']) {
+	
+	if (process.env['HOME']) {
+		process.env['ZIGBEE2MQTT_DATA'] = process.env['ZIGBEE2MQTT_DATA'] || path.join(process.env['HOME'], '.z2m');
```
Comment (sjorge) [inline cli.js:7]: This can be simplified as well, given we know the ENV var is unset.

Comment (Christian-Schl) [reply cli.js]: Done was just an copy and paste error

**Comment by Christian-Schl** [comment]
> 

@Koenkk can you please explain what kind of documentation your are expecting

**Comment by Koenkk** [comment]
@Christian-Schl I would expect it to be listed under https://www.zigbee2mqtt.io/guide/installation/

**Comment by Christian-Schl** [comment]
@Koenkk 
I not sure i an documentation for the end user makes sense if the end user should be getting benefits from the feature you should publish the binary with each release otherwise the documentation will look like this

- install nodejs
- clone git repo
- execute npm run build-bin
- start zigbee2mqtt using the binary. 

But this workflow make no sense because you need to download everything execute the build to create an independent binary.

If you publish the binary the installation will just look like this

- Download the binary for latest release
- Apply the config as documented https://www.zigbee2mqtt.io/guide/installation/04_windows.html#configuring


the update will be much easier 

- just download the latest binary and replace the current binary

**Comment by Christian-Schl** [comment]
@Koenkk 
now i am a little bit confused. Are you planning to add the executable to the next release ?

if yes sure we can update the documentation if not it makes no sense update the docs

**Comment by Koenkk** [comment]
@Christian-Schl 

Lets say we don't provide executables, what would be the use-case of this feature?

**Comment by nurikk** [comment]
If this feature can simplify windows users installation experience it worth trying.

**Comment by Christian-Schl** [comment]
@Koenkk 
if you dont provide executables. The setup will still be as it is. You need to setup the whole nodeJs env on every maschine where you what to use zigbee2mqtt. 

The thing what i want to optimize is 

- dont install the whole nodejs env on every machine where i want to use zigbee2mqtt
- simplify update just replease one file


You are the owner of the project so please decide if this make sense for you, if not just close the pull request also fine for me.

@nurikk 
This install experience  will work for all operation system (windows, linux, macos)

**Comment by Koenkk** [comment]
@Christian-Schl I'm not against this, what you mention is valid use case to me. However I still think some small docs need to be written. Maybe add an "Alternative installation" section to https://www.zigbee2mqtt.io/guide/installation/04_windows.html#starting-zigbee2mqtt and explain how to build the pkg there.

**Comment by nurikk** [comment]
> @Christian-Schl I'm not against this, what you mention is valid use case to me. However I still think some small docs need to be written. Maybe add an "Alternative installation" section to https://www.zigbee2mqtt.io/guide/installation/04_windows.html#starting-zigbee2mqtt and explain how to build the pkg there.

If I understand correctly we would need to add build pkg step into out ci/cd pipeline. And after execution it will generate .exe/linux binary which we will attach to github release and users free to download this binary file and do “double click installation”

**Comment by Christian-Schl** [comment]
@Koenkk 
sure if we commit to publish also the binary i can do this and wirte an documentation "Alternative installation"

Maybe also and it make sense to create also an documentation for the developer. How pkg is working and what are the importent thing. Maybe you can also point me here to an valid location

**Comment by nurikk** [comment]
I've checked this PR, and indeed, this pkg feature creates a standalone binary which can be executed without additional installation



```
➜  zigbee2mqtt git:(CS-External/master) npm run build-bin

> zigbee2mqtt@1.22.2 build-bin /Users/nur/code/zigbee2mqtt
> npm run build && pkg . --out-path bin --compress GZip


> zigbee2mqtt@1.22.2 build /Users/nur/code/zigbee2mqtt
> tsc && node index.js writehash

> pkg@5.5.2
compression:  GZip
> Targets not specified. Assuming:
  node14-linux-x64, node14-macos-x64, node14-win-x64
> Fetching base Node.js binaries to PKG_CACHE_PATH
  fetched-v14.18.3-linux-x64          [====================] 100%

  fetched-v14.18.3-macos-x64          [====================] 100%

  fetched-v14.18.3-win-x64            [====================] 100%

> Warning Cannot resolve ''../..' + '/package.json''
  /Users/nur/code/zigbee2mqtt/dist/util/utils.js
  Dynamic require may fail at run time, because the requested file
  is unknown at compilation time and not included into executable.
  Use a string literal as an argument for 'require', or leave it
  as is and specify the resolved file name in 'scripts' option.
> Warning Cannot resolve 'path_1.default.join(__dirname, '..', '..', 'node_modules', depend, 'package.json')'
  /Users/nur/code/zigbee2mqtt/dist/util/utils.js
  Dynamic require may fail at run time, because the requested file
  is unknown at compilation time and not included into executable.
  Use a string literal as an argument for 'require', or leave it
  as is and specify the resolved file name in 'scripts' option.
> Warning Cannot resolve 'moduleName'
  /Users/nur/code/zigbee2mqtt/dist/util/utils.js
  Dynamic require may fail at run time, because the requested file
  is unknown at compilation time and not included into executable.
  Use a string literal as an argument for 'require', or leave it
  as is and specify the resolved file name in 'scripts' option.
➜  zigbee2mqtt git:(CS-External/master) ls bin
 zigbee2mqtt-linux   zigbee2mqtt-macos   zigbee2mqtt-win.exe
➜  zigbee2mqtt git:(CS-External/master) cd bin
➜  bin git:(CS-External/master) ./zigbee2mqtt-macos
DataDir /Users/nur/.z2m
Zigbee2MQTT:info  2022-01-30 20:28:08: Logging to console only'
Zigbee2MQTT:info  2022-01-30 20:28:08: Starting Zigbee2MQTT version 1.22.2 (commit #6a8bbeac)
Zigbee2MQTT:info  2022-01-30 20:28:08: Starting zigbee-herdsman (0.13.188)
Zigbee2MQTT:error 2022-01-30 20:28:08: Error while starting zigbee-herdsman
Zigbee2MQTT:error 2022-01-30 20:28:08: Failed to start zigbee
Zigbee2MQTT:error 2022-01-30 20:28:08: Check https://www.zigbee2mqtt.io/guide/installation/20_zigbee2mqtt-fails-to-start.html for possible solutions
Zigbee2MQTT:error 2022-01-30 20:28:08: Exiting...
Zigbee2MQTT:error 2022-01-30 20:28:08: Error: Error while opening serialport 'Error: Error: No such file or directory, cannot open /dev/cu.usbserial-1140'
    at SerialPort.<anonymous> (/snapshot/zigbee2mqtt/node_modules/zigbee-herdsman/dist/adapter/z-stack/znp/znp.js:124:28)
    at SerialPort._error (/snapshot/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/stream/lib/index.js:198:14)
    at /snapshot/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/stream/lib/index.js:242:12
```

**Comment by nurikk** [comment]
To get it merged I guess we need
1) Update [ci/cd](https://github.com/Koenkk/zigbee2mqtt/blob/6ea61bae2cf441e15612dc68c257797fba46312b/.github/workflows/ci.yml#L28) pipeline to enable building binaries
2) [Attach](https://github.com/actions/upload-artifact) artefacts to GitHub release

File: cli.js [hunk 3]
```diff
@@ -1,4 +1,14 @@
 #!/usr/bin/env node
 const path = require('path');
-process.env['ZIGBEE2MQTT_DATA'] = process.env['ZIGBEE2MQTT_DATA'] || path.join(process.env['HOME'], '.z2m');
+
+if (!process.env['ZIGBEE2MQTT_DATA']) {
+	
+	if (process.env['HOME']) {
+		process.env['ZIGBEE2MQTT_DATA'] = path.join(process.env['HOME'], '.z2m');
```
Comment (21stce) [inline cli.js:7]: > $XDG_CONFIG_HOME defines the base directory relative to which user-specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.

**Comment by createcandle** [comment]
Just wanted to say that this sounds pretty useful to me, and would help my use case where I provide an addon for the Webthings Gateway. 
- Currently the addon downloads Z2M from github and installs it, which takes a while. If I could pre-build a binary and include that with the addon, that might save some local compile time. 
- It would also allow me to update the system Node environment without running into issues where Z2M then stops working correctly.

**Comment by 21stce** [comment]
@createcandle If you would base your work on NixOS, no such problem would exist.

**Comment by createcandle** [comment]
> @createcandle If you would base your work on NixOS, no such problem would exist.

:-D

**Comment by 21stce** [comment]
@mweinelt If you have a problem, you can just say it. I know you are perfectly capable of writing English. Using an emoticon implies a level of arrogance beyond Jupiter.

**Comment by createcandle** [comment]
> @mweinelt If you have a problem, you can just say it. I know you are perfectly capable of writing English. Using an emoticon implies a level of arrogance beyond Jupiter.

Maybe telling someone to rebase their project on a different OS without knowing the project (which already is a highly optimised OS build) is a little arrogant? While I'm sure you meant well, my reaction was to laugh, as this is so clearly not an option for me.

It's like me saying "I have some trouble with my boss", and you suggesting that the issue would be solved if I moved to a different country. While the suggestion is technically correct, that's all it is.

More generally, we're talking about making an install less dependent on the underlying OS here, so suggesting that switching to a specific OS is the solution is... odd?

Again, I do appreciate that you meant well, so my apologies if I was a little aloof.

**Comment by mweinelt** [comment]
I'm one of the NixOS maintainers for zigbee2mqtt, and even from my point of view that comment didn't feel well-placed. Fully agree with what @createcandle wrote.

**Comment by 21stce** [comment]
@mweinelt 

> I'm one of the NixOS maintainers for zigbee2mqtt, and even from my point of view

You are again suggesting that your point of view is somehow more informed. You are suggesting as if the fact that you maintain zigbee2mqtt means *anything*. 

@createcandle The fact that you aren't using NixOS or Guix as your base implies that you either believe you know better than the guy who wrote a PhD on the topic or that you are ignorant about the research. I am going to go with the hypothesis that you are ignorant, until otherwise proven. Indeed, I was just trying to help you avoid inventing a square wheel. Not a lot of people can create better wheels outside of NASA and Continental.

**Comment by 21stce** [comment]
@grahamc Can you please remove your ignorant face from the Internet?

**Comment by 21stce** [comment]
If the hate army has something to say that shows you have a combined IQ of over 150, then I am willing to listen, no problem, but all you fucking emoticon loving morons, please go fuck an ant.

**Comment by 21stce** [comment]
@grahamc Do you even have time left to do anything with all the harassing of people that you do?

**Comment by grahamc** [comment]
For the context of other people involved in this PR and repo, I just instituted a 7 day ban against @21stce on the NixOS organization for a pattern of abusive behavior. re: https://github.com/NixOS/nixpkgs/issues/164864#issuecomment-1090633821

**Comment by 21stce** [comment]
@mweinelt

Instead of keeping busy with emoticon discussions, perhaps you can fix what you likely have broken in unstable first:

```
 nixos systemd[1]: /etc/systemd/system/zigbee2mqtt.service:14: Invalid device node path 'tcp:/192.168.1.5:7777', ignoring.
```

That file contains on line 14:
```
DeviceAllow=tcp://192.168.1.5:7777
```

As such the error message is completely wrong too, which might be a bug in systemd. Anyway, but you wanted to play maintainer, so go *ahead*. Please notice that the number of `/` characters is *not* the same.

**Comment by wmertens** [comment]
Hi @21stce, maybe if this doesn't come from a maintainer it is easier to accept: your communication style comes over as very aggressive and that is unhelpful. I think we all want bug free things that do everything we need, so if you can just assume that, I'm sure you could be helpful too.

**Comment by sjorge** [comment]
@Koenkk seems this PR discussion has been thoroughly derailed, and some comments are probably not so inline with our Code of Conduct. Perhaps the PR should be closed or some comments hidden.

**Comment by rohankapoorcom** [comment]
@Koenkk @Christian-Schl I'm new to zigbee2mqtt but very much interested in the feature. Looks like this one got stuck. What else is needed to move this forward? I'm happy to assist.

**Pull request closed without merging**